### DISCUSSION TITLE
How to read x-data object from outside of Alpine

### DISCUSSION THREAD

**Discussion by Lomacar** [alpinejs/alpine#2375] [1. Help] [answered]
I'm trying to make a js component system that works alongside Alpine. I have module code attached to each component and actually defined inside the component template in a script tag.  Like this:
```html
<template id="test-component">
    <script>
        export const foo = "bar"
   </script>
  <p>Content goes here.</p>
</template>
```
My system extracts the script tag and converts it into a module object attached to each custom element instance. So anyway, I want to be able to interact with Alpine, so that I could supply some `x-data` like this:
```html
<test-component x-data="{foo: 'baz'}">
```
And then, within the component I could say something like:
```javascript
export const foo = xdata.foo
```
But how could I get xdata to point to the x-data? By the way, I have it set up so that a variable called `host` refers to the top node of the component, so I can say `host.getAttribute('x-data')` but of course that only gets me the string. I can even do `Alpine.closestRoot(host)`, but that is just the same as `host` by itself. I was hoping that I might be able to use Alpine.data(host) or something, but it doesn't work that way.

**Comment by SimoTod** [comment, marked as answer]
There's no public api to get the data context from the object. However I believe `Alpine.mergeProxies(host._x_dataStack)` should be what you are after

**Reply by Lomacar** [reply]
Wow, thank you. At first I thought mergeProxies was doing nothing, and I wasted a bunch of time trying to do my own thing with _x_dataStack, and then I realized that even though the result of mergeProxies looks obscure, you can still just refer to all your x-data properties on it.

**Reply by Lomacar** [reply]
It looks like `Alpine.evaluate(host, '$data')` does the same thing. Any difference?

**Reply by Lomacar** [reply]
Actually, I am noticing that when I try to access something undefined from the data stack with mergeProxies I get an uncaught error `objects is undefined`, but with the evaluate method I get no error.

**Reply by SimoTod** [reply]
Yeah, you are probably better off using $data which is the official api (it's a recent addition so I didn't think about that at first). It does almost the same thing (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/magics/%24data.js#L5).
I believe the error is captured by evaluate, in your case.
